### PR TITLE
Implement Transformer Decoder and Encoder-Decoder Model

### PR DESCRIPTION
**Pull request by JacobBorden**
This PR adds the missing Decoder component to the Transformer module. It implements `TransformerDecoderLayer` with masked self-attention, cross-attention, and an FFN. It also introduces `TransformerEncoderDecoderModel` to tie the entire sequence-to-sequence structure together. The build files and TODO lists were updated, and tests for the new architecture were written to verify forward pass operation.

PR steward update for 2026-06-30:
- Rebased onto current `master` after #136 merged.
- Added default causal decoder self-attention mask generation when `forward(src, tgt)` omits `tgt_self_attention_mask`.
- Added regression coverage that future target-token changes do not affect earlier decoder outputs under the default mask.
- Local validation passed with Python 3.11 pinned: `cmake -B build -DCMAKE_BUILD_TYPE=Release -DPython3_EXECUTABLE=/bin/python3.11 -DPYTHON_EXECUTABLE=/bin/python3.11`, `cmake --build build --config Release`, and `ctest --test-dir build -C Release --output-on-failure` (178/178).

---
*PR created automatically by Jules for task [9930568919225893395](https://jules.google.com/task/9930568919225893395) started by @JacobBorden*

### CODE REVIEW

File: src/transformer/transformer_encoder_decoder_model.h [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+     * @param tgt_self_attention_mask Optional mask for decoder self-attention.
+     * @param tgt_cross_attention_mask Optional mask for decoder cross-attention.
+     * @return Matrix::Matrix<float> Output matrix from the final decoder layer,
+     *                               after final layer normalization. Shape (tgt_seq_len, d_model).
+     */
+    Matrix::Matrix<float> forward(
+        const Matrix::Matrix<float>& src_input_token_ids,
+        const Matrix::Matrix<float>& tgt_input_token_ids,
+        const Matrix::Matrix<float>& src_attention_mask = Matrix::Matrix<float>(0,0),
+        const Matrix::Matrix<float>& tgt_self_attention_mask = Matrix::Matrix<float>(0,0),
```
[the remark this answers is not in the document]
Comment (JacobBorden) [reply src/transformer/transformer_encoder_decoder_model.h:58]: Fixed in `2e135b6`: the encoder-decoder forward path now generates an additive causal decoder self-attention mask whenever `tgt_self_attention_mask` is omitted, and `DefaultDecoderMaskPreventsFutureTokenLeakage` covers the regression by changing a future target token and asserting earlier decoder outputs remain unchanged. Local Release build and CTest passed with 178/178 tests.
